File: compatibility-tests/sdk-test-go/tests/s3_notifications_test.go
package tests

import (
	"context"
	"testing"

	"floci-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/aws/aws-sdk-go-v2/service/sns"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestS3Notifications(t *testing.T) {
	ctx := context.Background()
	s3Svc := testutil.S3Client()
	sqsSvc := testutil.SQSClient()
	snsSvc := testutil.SNSClient()

	queueName := "s3-notif-filter-queue"
	topicName := "s3-notif-filter-topic"
	bucketName := "s3-notif-filter-bucket"
	queueArn := "arn:aws:sqs:us-east-1:000000000000:" + queueName
	var topicArn string
	var queueURL string

	t.Cleanup(func() {
		s3Svc.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucketName)})
		if queueURL != "" {
			sqsSvc.DeleteQueue(ctx, &sqs.DeleteQueueInput{QueueUrl: aws.String(queueURL)})
		}
		if topicArn != "" {
			snsSvc.DeleteTopic(ctx, &sns.DeleteTopicInput{TopicArn: aws.String(topicArn)})
		}
	})

	t.Run("Setup", func(t *testing.T) {
		// Create SQS queue
		qr, err := sqsSvc.CreateQueue(ctx, &sqs.CreateQueueInput{QueueName: aws.String(queueName)})
		require.NoError(t, err)
		queueURL = aws.ToString(qr.QueueUrl)

		// Create SNS topic
		tr, err := snsSvc.CreateTopic(ctx, &sns.CreateTopicInput{Name: aws.String(topicName)})
		require.NoError(t, err)
		topicArn = aws.ToString(tr.TopicArn)

		// Create S3 bucket
		_, err = s3Svc.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucketName)})
		require.NoError(t, err)
	})

	t.Run("PutBucketNotificationConfiguration", func(t *testing.T) {
		_, err := s3Svc.PutBucketNotificationConfiguration(ctx, &s3.PutBucketNotificationConfigurationInput{
			Bucket: aws.String(bucketName),
			NotificationConfiguration: &s3types.NotificationConfiguration{
				QueueConfigurations: []s3types.QueueConfiguration{
					{
						Id:       aws.String("sqs-filtered"),
						QueueArn: aws.String(queueArn),
						Events:   []s3types.Event{s3types.EventS3ObjectCreated},
						Filter: &s3types.NotificationConfigurationFilter{
							Key: &s3types.S3KeyFilter{
								FilterRules: []s3types.FilterRule{
									{Name: s3types.FilterRuleNamePrefix, Value: aws.String("incoming/")},
									{Name: s3types.FilterRuleNameSuffix, Value: aws.String(".csv")},
								},
							},
						},
					},
				},
				TopicConfigurations: []s3types.TopicConfiguration{
					{
						Id:       aws.String("sns-filtered"),
						TopicArn: aws.String(topicArn),
						Events:   []s3types.Event{s3types.EventS3ObjectRemoved},
						Filter: &s3types.NotificationConfigurationFilter{
							Key: &s3types.S3KeyFilter{
								FilterRules: []s3types.FilterRule{
									{Name: s3types.FilterRuleNamePrefix, Value: aws.String("")},
									{Name: s3types.FilterRuleNameSuffix, Value: aws.String(".txt")},
								},
							},
						},
					},
				},
			},
		})
		require.NoError(t, err)
	})

	t.Run("GetBucketNotificationConfiguration", func(t *testing.T) {
		r, err := s3Svc.GetBucketNotificationConfiguration(ctx, &s3.GetBucketNotificationConfigurationInput{
			Bucket: aws.String(bucketName),
		})
		require.NoError(t, err)

		// Assert queue config
		require.NotEmpty(t, r.QueueConfigurations)
		assert.Equal(t, queueArn, aws.ToString(r.QueueConfigurations[0].QueueArn))
		require.NotNil(t, r.QueueConfigurations[0].Filter)
		require.NotNil(t, r.QueueConfigurations[0].Filter.Key)
		assert.Len(t, r.QueueConfigurations[0].Filter.Key.FilterRules, 2)

		// Assert topic config
		require.NotEmpty(t, r.TopicConfigurations)
		assert.Equal(t, topicArn, aws.ToString(r.TopicConfigurations[0].TopicArn))
		require.NotNil(t, r.TopicConfigurations[0].Filter)
		require.NotNil(t, r.TopicConfigurations[0].Filter.Key)
		assert.Len(t, r.TopicConfigurations[0].Filter.Key.FilterRules, 2)
	})
}

File: compatibility-tests/sdk-test-rust/tests/s3_notifications_test.rs
//! S3 Bucket Notification Configuration Tests
//!
//! Tests S3 notification configurations with SQS and SNS,
//! including filter rules for prefix/suffix patterns.

mod common;

use aws_sdk_s3::types::{
    Event as S3Event, FilterRule as S3FilterRule, FilterRuleName,
    NotificationConfiguration, NotificationConfigurationFilter,
    QueueConfiguration as S3QueueConfig, S3KeyFilter,
    TopicConfiguration as S3TopicConfig,
};

#[tokio::test(flavor = "multi_thread")]
async fn test_s3_notifications_with_filters() {
    let s3 = common::s3_client().await;
    let sqs = common::sqs_client().await;
    let sns = common::sns_client().await;
    let endpoint = sdk_test_rust::endpoint();

    let bucket = "rust-s3-notif-filter-bucket";
    let queue_name = "rust-s3-notif-filter-queue";
    let topic_name = "rust-s3-notif-filter-topic";
    let account_id = "000000000000";
    let queue_arn = format!("arn:aws:sqs:us-east-1:{}:{}", account_id, queue_name);

    // Create topic first to get ARN
    let topic_result = sns.create_topic().name(topic_name).send().await;
    let topic_arn = match topic_result {
        Ok(r) => r.topic_arn.unwrap_or_default(),
        Err(e) => panic!("Failed to create SNS topic: {:?}", e),
    };

    let _guard = common::CleanupGuard::new({
        let s3 = s3.clone();
        let sqs = sqs.clone();
        let sns = sns.clone();
        let endpoint = endpoint.clone();
        let topic_arn = topic_arn.clone();
        async move {
            let _ = s3.delete_bucket().bucket(bucket).send().await;
            let queue_url = format!("{}/{}/{}", endpoint, account_id, queue_name);
            let _ = sqs.delete_queue().queue_url(&queue_url).send().await;
            let _ = sns.delete_topic().topic_arn(&topic_arn).send().await;
        }
    });

    // Create SQS queue
    sqs.create_queue()
        .queue_name(queue_name)
        .send()
        .await
        .expect("create queue");

    // Create S3 bucket
    s3.create_bucket()
        .bucket(bucket)
        .send()
        .await
        .expect("create bucket");

    // Configure queue notification with prefix/suffix filter
    let queue_config = S3QueueConfig::builder()
        .id("sqs-filtered")
        .queue_arn(&queue_arn)
        .events(S3Event::S3ObjectCreated)
        .filter(
            NotificationConfigurationFilter::builder()
                .key(
                    S3KeyFilter::builder()
                        .filter_rules(
                            S3FilterRule::builder()
                                .name(FilterRuleName::Prefix)
                                .value("incoming/")
                                .build(),
                        )
                        .filter_rules(
                            S3FilterRule::builder()
                                .name(FilterRuleName::Suffix)
                                .value(".csv")
                                .build(),
                        )
                        .build(),
                )
                .build(),
        )
        .build()
        .expect("queue config");

    // Configure topic notification with suffix filter
    let topic_config = S3TopicConfig::builder()
        .id("sns-filtered")
        .topic_arn(&topic_arn)
        .events(S3Event::S3ObjectRemoved)
        .filter(
            NotificationConfigurationFilter::builder()
                .key(
                    S3KeyFilter::builder()
                        .filter_rules(
                            S3FilterRule::builder()
                                .name(FilterRuleName::Prefix)
                                .value("")
                                .build(),
                        )
                        .filter_rules(
                            S3FilterRule::builder()
                                .name(FilterRuleName::Suffix)
                                .value(".txt")
                                .build(),
                        )
                        .build(),
                )
                .build(),
        )
        .build()
        .expect("topic config");

    // Put notification configuration
    let put_result = s3
        .put_bucket_notification_configuration()
        .bucket(bucket)
        .notification_configuration(
            NotificationConfiguration::builder()
                .queue_configurations(queue_config)
                .topic_configurations(topic_config)
                .build(),
        )
        .send()
        .await;

    assert!(
        put_result.is_ok(),
        "PutBucketNotificationConfiguration failed: {:?}",
        put_result.err()
    );

    // Get and verify notification configuration
    let get_result = s3
        .get_bucket_notification_configuration()
        .bucket(bucket)
        .send()
        .await;

    assert!(
        get_result.is_ok(),
        "GetBucketNotificationConfiguration failed: {:?}",
        get_result.err()
    );

    let nc = get_result.unwrap();

    // Verify queue configuration
    let queue_configs = nc.queue_configurations();
    assert!(
        !queue_configs.is_empty(),
        "should have queue configurations"
    );
    let has_queue = queue_configs.iter().any(|q| q.queue_arn() == queue_arn);
    assert!(has_queue, "queue ARN should match");

    // Verify queue filter rules
    let queue_filter_ok = queue_configs.iter().any(|q| {
        q.queue_arn() == queue_arn
            && q.filter()
                .and_then(|f| f.key())
                .map(|k| k.filter_rules().len() == 2)
                .unwrap_or(false)
    });
    assert!(
        queue_filter_ok,
        "queue should have 2 filter rules (prefix and suffix)"
    );

    // Verify topic configuration
    let topic_configs = nc.topic_configurations();
    assert!(
        !topic_configs.is_empty(),
        "should have topic configurations"
    );
    let has_topic = topic_configs.iter().any(|t| t.topic_arn() == topic_arn);
    assert!(has_topic, "topic ARN should match");

    // Verify topic filter rules
    let topic_filter_ok = topic_configs.iter().any(|t| {
        t.topic_arn() == topic_arn
            && t.filter()
                .and_then(|f| f.key())
                .map(|k| k.filter_rules().len() == 2)
                .unwrap_or(false)
    });
    assert!(
        topic_filter_ok,
        "topic should have 2 filter rules (prefix and suffix)"
    );
}
